### PR TITLE
ENG-8608: Introduce a specialized update record to the DR binary log …

### PR DESCRIPTION
**Pull request by afrogers**
…format

### CODE REVIEW

File: src/frontend/org/voltdb/PartitionDRGateway.java [hunk 1]
```diff
@@ -183,6 +183,28 @@ public static synchronized void pushDRBuffer(
...
+                        oldRowIndexCrc = 0;
+                    }
+                    buf.position(buf.position() + oldRowLengthPrefix);
+                    final int newRowLengthPrefix = buf.getInt();
+                    buf.position(buf.position() + newRowLengthPrefix);
+                    checksum = buf.getInt();
+                    log.trace("Version " + version + " type " + recordType + " table handle " + tableHandle + " old row length " + oldRowLengthPrefix +
+                              " new row length " + newRowLengthPrefix + " checksum " + checksum +
+                              (recordType == DRRecordType.UPDATE_BY_INDEX ? (" index checksum " + oldRowIndexCrc) : ""));
+                }
```
Comment (slowsoul) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: missing break here

Comment (slowsoul) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: In truncateTable() implementation in DRTupleStream.cpp, it is better to use transactionChecks() and TXN_RECORD_HEADER_SIZE which have been factored out and used in the other two: appendTuple() and appendUpdateRecord().

Comment (slowsoul) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: In line 678 of DRBinaryLog_test.cpp, the result should be assigned to temp_tuple, although current code does not affect correctness.

Comment (slowsoul) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: In Test **FillSingleTxnAndAppend** and **FillSingleTxnAndFlush** of DRTupleStream_test.cpp, the comments regarding the value of **m_topend.receivedDRBuffer** seem to be outdated/inconsistent.

File: src/ee/storage/BinaryLogSink.cpp [hunk 2]
```diff
@@ -74,73 +75,159 @@ void BinaryLogSink::apply(const char *taskParams, boost::unordered_map<int64_t,
         pool->purge();
         const char* recordStart = taskInfo.getRawPointer();
         const uint8_t drVersion = taskInfo.readByte();
-        if (drVersion > 1) {
+        if (drVersion > 2) {
```
Comment (luyangco) [inline src/ee/storage/BinaryLogSink.cpp]: Should backward compatibility be considered here?

**Pull request closed without merging**